packages/spec/src/v1.ts

```diff
@@ -80,15 +87,15 @@ export interface StandardIssue {
   /**
    * The path of the issue, if any.
    */
-  readonly path?: ReadonlyArray<PropertyKey | StandardPathItem> | undefined;
+  readonly path?: ReadonlyArray<PropertyKey | StandardPathSegment> | undefined;
 }
 
 /**
  * The path item interface of the issue.
  */
-export interface StandardPathItem {
+export interface StandardPathSegment {
   /**
-   * The key of the path item.
+   * The key of the path segment.
    */
-  readonly key: unknown;
+  readonly key: PropertyKey;
```

Comment:
Presumably, this was typed as `unknown` for flexibility, but as an external consumer, I have no idea what I would do with `key: unknown`.

I think this would be much more externally useful if we also narrowed it to `PropertyKey`. If there are use cases for other kinds of paths that wouldn't support this, maybe they could be specified via a different API that would make their intent clear?

Reply:
Valibot always returns an object with a `key` property instead of just a `PropertyKey` like Zod, and to be consistent throughout the library, `key` will always be the actual `key` of the received input. Since `Map` supports any value for its key, `PropertyKey` would break Valibots compatibility. Developers can still check if the unknown key is of type `PropertyKey`, as shown here: https://github.com/standard-schema/standard-schema/blob/main/packages/utils/src/getDotPath/getDotPath.ts#L15

So again, this is just for maximum flexibility. 🤪 How does ArkType handle this? Is `path` always an array or a tuple of `PropertyKey`?

Reply:
Ahh I see, that makes sense. Yeah, ArkType's concept of keys/paths mirrors that of JS, so if you were traversing into a map value to validate, the path would not go beyond the root value of the Map itself, and additional details would have to be added to the error there.

I would guess this use case of wanting a path including e.g. an object or similar from a Map would be pretty rare, to the point perhaps there's more value in just only allowing `PropertyKey` and doing something similar where the path is truncated where the `Map` starts?

I see how you can still narrow it as a consumer where needed, but I do think we're giving up a lot to have a type that is widened to the point it's not clear what it's supposed to represent or immediately obvious what to do with it if you're just looking for the 99% case a `PropertyKey`. Even working on validation all the time, I couldn't figure out that a non-`PropertyKey` value was supposed to represent a `Map` entry, and even after knowing it, I think I'd struggle to know what do with e.g. an object or function as a library author trying to represent a path.


Another option would be to have a second optional path key like `mappablePath` or similar that widened the key type to `unknown` and could be provided by validators that support it alongside the standard `PropertyKey` constrained `path`. This would still give library authors a straightforward way to handle the cases they're probably interested in, while allowing that extra information to be available in cases where it's useful.

Reply:
Our approach with the current spec was to make it as flexible as possible to get everyone on board while keeping the spec simple to understand and implement. I think both path key approaches are valid, but changing the spec in this regard would force Valibot to compromise on its consistency and bundle size.

Reply:
I definitely agree on the goals, and it was definitely very straightforward to add support for this in ArkType!

However, most of my concerns are to do with the cost of that simplicity and flexibility in terms of impact on consumers.

Every edge case or alternative we add here makes it easier for a validator to satisfy the API, but simultaneously is one more case that consumers have to specifically account for.

IMO there is relatively little downside to adding optional properties for additional introspectability (e.g. `mappablePath`), but the cost of supporting multiple path formats is that those formats have to be accounted for by every would-be consumer.

This is particularly problematic in a case like this one, where it may not be at all clear what a consumer should even do with an object like `{ key: unknown }` if they are expecting a simple path. If anything, the path should probably just be `PropertyKey[]` so it's as easy to consume as possible.

I get that it may make satisfying the spec slightly less optimal in terms of bundle size if it doesn't perfectly align with what you've already implemented, but I think for this standard to be successful, there has to be willingness for individual validators to align to the simplest possible representation that most people will be expecting. If there are too many edge cases to handle because different validators want a spec that aligns with their pre-existing formats, devs considering accepting a Standard Schema in their API could just throw their hands up and support only a single validator directly because the DX gap between that validator's API and that of Standard Schema is too large.